File: src/lib/workflow/schemas.ts
/**
 * Client-side mirror of the FastAPI workflow contract.
 *
 * The backend in `backend/app/workflow/schemas.py` is the source of truth; these schemas
 * validate responses at the network boundary and provide types for the UI.
 */
import { z } from "zod";

export const SeveritySchema = z.enum(["SEV-1", "SEV-2", "SEV-3"]);
export type Severity = z.infer<typeof SeveritySchema>;

export const TeamSchema = z.enum([
  "command",
  "triage",
  "investigation",
  "response",
  "communications",
]);
export type Team = z.infer<typeof TeamSchema>;

export const AgentRoleSchema = z.enum([
  "main-orchestrator",
  "sub-orchestrator",
  "specialist",
]);
export type AgentRole = z.infer<typeof AgentRoleSchema>;

export const AgentStatusSchema = z.enum([
  "queued",
  "running",
  "completed",
  "blocked",
  "cancelled",
  "failed",
]);
export type AgentStatus = z.infer<typeof AgentStatusSchema>;

export const WorkflowPhaseSchema = z.enum([
  "intake",
  "triage",
  "investigation",
  "planning",
  "approval",
  "remediation",
  "verification",
  "resolved",
  "rejected",
  "failed",
]);
export type WorkflowPhase = z.infer<typeof WorkflowPhaseSchema>;

export const WorkflowStatusSchema = z.enum([
  "running",
  "awaiting_approval",
  "completed",
  "rejected",
  "failed",
]);
export type WorkflowStatus = z.infer<typeof WorkflowStatusSchema>;

export const IncidentInputSchema = z
  .object({
    title: z.string().trim().min(5).max(120),
    description: z.string().trim().min(20).max(2_000),
    service: z
      .string()
      .trim()
      .min(2)
      .max(80)
      .regex(/^[a-zA-Z0-9._/-]+$/, "Use a service name without spaces"),
    severity: SeveritySchema,
    region: z.string().trim().min(2).max(80),
    signals: z.array(z.string().trim().min(3).max(240)).min(1).max(12),
  })
  .strict();
export type IncidentInput = z.infer<typeof IncidentInputSchema>;

export const AgentRuntimeSchema = z
  .object({
    id: z.string().min(1),
    name: z.string().min(1),
    shortName: z.string().min(1),
    role: AgentRoleSchema,
    team: TeamSchema,
    parentId: z.string().nullable(),
    mission: z.string().min(1),
    capabilities: z.array(z.string().min(1)).min(1),
    status: AgentStatusSchema,
    currentTask: z.string().nullable(),
    outputSummary: z.string().nullable(),
    startedAt: z.string().datetime().nullable(),
    completedAt: z.string().datetime().nullable(),
  })
  .strict();
export type AgentRuntime = z.infer<typeof AgentRuntimeSchema>;

export const FindingSeveritySchema = z.enum(["info", "warning", "critical"]);

export const AgentFindingSchema = z
  .object({
    id: z.string().min(1),
    agentId: z.string().min(1),
    team: TeamSchema,
    headline: z.string().min(3).max(160),
    detail: z.string().min(10).max(1_200),
    evidence: z.array(z.string().min(3).max(320)).min(1).max(8),
    confidence: z.number().min(0).max(1),
    severity: FindingSeveritySchema,
  })
  .strict();
export type AgentFinding = z.infer<typeof AgentFindingSchema>;

export const TeamReportSchema = z
  .object({
    id: z.string().min(1),
    orchestratorId: z.string().min(1),
    team: TeamSchema,
    title: z.string().min(3).max(160),
    summary: z.string().min(10).max(1_500),
    keyFindings: z.array(z.string().min(3).max(300)).min(1).max(8),
    recommendation: z.string().min(5).max(800),
    confidence: z.number().min(0).max(1),
  })
  .strict();
export type TeamReport = z.infer<typeof TeamReportSchema>;

export const RiskLevelSchema = z.enum(["low", "medium", "high"]);

export const RemediationActionSchema = z
  .object({
    id: z.string().min(1),
    title: z.string().min(3).max(160),
    detail: z.string().min(10).max(800),
    ownerAgentId: z.string().min(1),
    risk: RiskLevelSchema,
    reversible: z.boolean(),
    expectedSignal: z.string().min(5).max(400),
  })
  .strict();
export type RemediationAction = z.infer<typeof RemediationActionSchema>;

export const RemediationPlanSchema = z
  .object({
    id: z.string().min(1),
    hypothesis: z.string().min(10).max(1_000),
    summary: z.string().min(10).max(1_000),
    riskLevel: RiskLevelSchema,
    blastRadius: z.string().min(5).max(500),
    actions: z.array(RemediationActionSchema).min(1).max(8),
    rollback: z.string().min(10).max(800),
    validationChecks: z.array(z.string().min(5).max(320)).min(1).max(10),
  })
  .strict();
export type RemediationPlan = z.infer<typeof RemediationPlanSchema>;

export const ApprovalRequestSchema = z
  .object({
    id: z.string().min(1),
    status: z.enum(["pending", "approved", "rejected"]),
    requestedAt: z.string().datetime(),
    decidedAt: z.string().datetime().nullable(),
    decidedBy: z.string().nullable(),
    note: z.string().nullable(),
    plan: RemediationPlanSchema,
  })
  .strict();
export type ApprovalRequest = z.infer<typeof ApprovalRequestSchema>;

export const VerificationCheckSchema = z
  .object({
    label: z.string().min(3).max(160),
    value: z.string().min(1).max(120),
    status: z.enum(["passed", "failed"]),
    detail: z.string().min(3).max(400),
  })
  .strict();

export const VerificationReportSchema = z
  .object({
    status: z.enum(["passed", "failed"]),
    summary: z.string().min(10).max(1_000),
    checks: z.array(VerificationCheckSchema).min(1).max(10),
  })
  .strict();
export type VerificationReport = z.infer<typeof VerificationReportSchema>;

export const IncidentOutcomeSchema = z
  .object({
    rootCause: z.string().min(10).max(1_000),
    resolution: z.string().min(10).max(1_000),
    customerImpact: z.string().min(10).max(800),
    followUps: z.array(z.string().min(5).max(320)).min(1).max(10),
  })
  .strict();
export type IncidentOutcome = z.infer<typeof IncidentOutcomeSchema>;

export const WorkflowEventTypeSchema = z.enum([
  "workflow-started",
  "delegation",
  "agent-started",
  "finding",
  "synthesis",
  "plan-ready",
  "approval-requested",
  "approval-granted",
  "approval-rejected",
  "remediation",
  "verification",
  "communication",
  "workflow-completed",
  "workflow-failed",
]);
export type WorkflowEventType = z.infer<typeof WorkflowEventTypeSchema>;

export const WorkflowEventSchema = z
  .object({
    id: z.string().min(1),
    sequence: z.number().int().positive(),
    timestamp: z.string().datetime(),
    type: WorkflowEventTypeSchema,
    phase: WorkflowPhaseSchema,
    actorId: z.string().min(1),
    actorName: z.string().min(1),
    team: TeamSchema,
    title: z.string().min(3).max(160),
    detail: z.string().min(3).max(1_200),
    level: z.enum(["neutral", "success", "warning", "critical"]),
  })
  .strict();
export type WorkflowEvent = z.infer<typeof WorkflowEventSchema>;

export const WorkflowMetricsSchema = z
  .object({
    agentsTotal: z.number().int().nonnegative(),
    agentsCompleted: z.number().int().nonnegative(),
    activeAgents: z.number().int().nonnegative(),
    tasksCompleted: z.number().int().nonnegative(),
    handoffs: z.number().int().nonnegative(),
    confidence: z.number().min(0).max(1),
  })
  .strict();
export type WorkflowMetrics = z.infer<typeof WorkflowMetricsSchema>;

export const WorkflowRunSchema = z
  .object({
    id: z.string().min(1),
    incident: IncidentInputSchema,
    mode: z.enum(["demo", "openai"]),
    status: WorkflowStatusSchema,
    phase: WorkflowPhaseSchema,
    startedAt: z.string().datetime(),
    updatedAt: z.string().datetime(),
    agents: z.array(AgentRuntimeSchema),
    events: z.array(WorkflowEventSchema),
    findings: z.array(AgentFindingSchema),
    teamReports: z.array(TeamReportSchema),
    plan: RemediationPlanSchema.nullable(),
    approval: ApprovalRequestSchema.nullable(),
    verification: VerificationReportSchema.nullable(),
    outcome: IncidentOutcomeSchema.nullable(),
    metrics: WorkflowMetricsSchema,
  })
  .strict();
export type WorkflowRun = z.infer<typeof WorkflowRunSchema>;

export const ApprovalDecisionSchema = z
  .object({
    decision: z.enum(["approve", "reject"]),
    reviewer: z.string().trim().min(2).max(80),
    note: z.string().trim().max(500).default(""),
  })
  .strict();
export type ApprovalDecision = z.infer<typeof ApprovalDecisionSchema>;

export const ApiErrorSchema = z.object({
  error: z.string(),
  details: z.array(z.string()).optional(),
});
export type ApiError = z.infer<typeof ApiErrorSchema>;

export const DEFAULT_INCIDENT: IncidentInput = {
  title: "Checkout latency surge across EU region",
  description:
    "Checkout p95 latency climbed from 420 ms to 8.4 s shortly after the payment-router deployment. Error rate is 18% in eu-central and the retry queue continues to grow.",
  service: "payment-router",
  severity: "SEV-1",
  region: "eu-central",
  signals: [
    "p95 latency 8.4 s (baseline 420 ms)",
    "HTTP 5xx rate 18%",
    "payment retry queue 4.7× above baseline",
  ],
};

File: backend/app/workflow/schemas.py
"""Runtime and transport contracts for the Twemp incident workflow.

The JSON contract is camelCase so the existing frontend consumes this API unchanged.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Annotated, Literal

from pydantic import (
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    PlainSerializer,
    StringConstraints,
)
from pydantic.alias_generators import to_camel

Severity = Literal["SEV-1", "SEV-2", "SEV-3"]
Team = Literal["command", "triage", "investigation", "response", "communications"]
OperationalTeam = Literal["triage", "investigation", "response", "communications"]
AgentRole = Literal["main-orchestrator", "sub-orchestrator", "specialist"]
AgentStatus = Literal["queued", "running", "completed", "blocked", "cancelled", "failed"]
FindingSeverity = Literal["info", "warning", "critical"]
RiskLevel = Literal["low", "medium", "high"]
EventLevel = Literal["neutral", "success", "warning", "critical"]
ApprovalStatus = Literal["pending", "approved", "rejected"]
CheckStatus = Literal["passed", "failed"]
ProviderMode = Literal["demo", "openai"]

WorkflowPhase = Literal[
    "intake",
    "triage",
    "investigation",
    "planning",
    "approval",
    "remediation",
    "verification",
    "resolved",
    "rejected",
    "failed",
]

WorkflowStatus = Literal["running", "awaiting_approval", "completed", "rejected", "failed"]

WorkflowEventType = Literal[
    "workflow-started",
    "delegation",
    "agent-started",
    "finding",
    "synthesis",
    "plan-ready",
    "approval-requested",
    "approval-granted",
    "approval-rejected",
    "remediation",
    "verification",
    "communication",
    "workflow-completed",
    "workflow-failed",
]


def utc_now() -> datetime:
    """Return the current time as a timezone-aware UTC value."""
    return datetime.now(UTC)


def _to_iso_z(value: datetime) -> str:
    """Serialize to the ISO-8601 `Z` form the frontend contract expects."""
    return value.astimezone(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")


Timestamp = Annotated[AwareDatetime, PlainSerializer(_to_iso_z, return_type=str, when_used="json")]

Identifier = Annotated[str, StringConstraints(min_length=1)]
Confidence = Annotated[float, Field(ge=0, le=1)]
Headline = Annotated[str, StringConstraints(min_length=3, max_length=160)]


class ContractModel(BaseModel):
    """Base model that serializes camelCase, rejects unknown keys, and validates mutations."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
        validate_assignment=True,
    )


Signal = Annotated[str, StringConstraints(strip_whitespace=True, min_length=3, max_length=240)]


class IncidentInput(ContractModel):
    title: Annotated[str, StringConstraints(strip_whitespace=True, min_length=5, max_length=120)]
    description: Annotated[
        str, StringConstraints(strip_whitespace=True, min_length=20, max_length=2000)
    ]
    service: Annotated[
        str,
        StringConstraints(
            strip_whitespace=True,
            min_length=2,
            max_length=80,
            pattern=r"^[A-Za-z0-9._/-]+$",
        ),
    ]
    severity: Severity
    region: Annotated[str, StringConstraints(strip_whitespace=True, min_length=2, max_length=80)]
    signals: Annotated[list[Signal], Field(min_length=1, max_length=12)]


class AgentRuntime(ContractModel):
    id: Identifier
    name: Identifier
    short_name: Identifier
    role: AgentRole
    team: Team
    parent_id: Identifier | None
    mission: Identifier
    capabilities: Annotated[list[Identifier], Field(min_length=1)]
    status: AgentStatus
    current_task: str | None
    output_summary: str | None
    started_at: Timestamp | None
    completed_at: Timestamp | None


class AgentFinding(ContractModel):
    id: Identifier
    agent_id: Identifier
    team: Team
    headline: Headline
    detail: Annotated[str, StringConstraints(min_length=10, max_length=1200)]
    evidence: Annotated[
        list[Annotated[str, StringConstraints(min_length=3, max_length=320)]],
        Field(min_length=1, max_length=8),
    ]
    confidence: Confidence
    severity: FindingSeverity


class TeamReport(ContractModel):
    id: Identifier
    orchestrator_id: Identifier
    team: Team
    title: Headline
    summary: Annotated[str, StringConstraints(min_length=10, max_length=1500)]
    key_findings: Annotated[
        list[Annotated[str, StringConstraints(min_length=3, max_length=300)]],
        Field(min_length=1, max_length=8),
    ]
    recommendation: Annotated[str, StringConstraints(min_length=5, max_length=800)]
    confidence: Confidence


class RemediationAction(ContractModel):
    id: Identifier
    title: Headline
    detail: Annotated[str, StringConstraints(min_length=10, max_length=800)]
    owner_agent_id: Identifier
    risk: RiskLevel
    reversible: bool
    expected_signal: Annotated[str, StringConstraints(min_length=5, max_length=400)]


class RemediationPlan(ContractModel):
    id: Identifier
    hypothesis: Annotated[str, StringConstraints(min_length=10, max_length=1000)]
    summary: Annotated[str, StringConstraints(min_length=10, max_length=1000)]
    risk_level: RiskLevel
    blast_radius: Annotated[str, StringConstraints(min_length=5, max_length=500)]
    actions: Annotated[list[RemediationAction], Field(min_length=1, max_length=8)]
    rollback: Annotated[str, StringConstraints(min_length=10, max_length=800)]
    validation_checks: Annotated[
        list[Annotated[str, StringConstraints(min_length=5, max_length=320)]],
        Field(min_length=1, max_length=10),
    ]


class ApprovalRequest(ContractModel):
    id: Identifier
    status: ApprovalStatus
    requested_at: Timestamp
    decided_at: Timestamp | None
    decided_by: str | None
    note: str | None
    plan: RemediationPlan


class VerificationCheck(ContractModel):
    label: Headline
    value: Annotated[str, StringConstraints(min_length=1, max_length=120)]
    status: CheckStatus
    detail: Annotated[str, StringConstraints(min_length=3, max_length=400)]


class VerificationReport(ContractModel):
    status: CheckStatus
    summary: Annotated[str, StringConstraints(min_length=10, max_length=1000)]
    checks: Annotated[list[VerificationCheck], Field(min_length=1, max_length=10)]


class IncidentOutcome(ContractModel):
    root_cause: Annotated[str, StringConstraints(min_length=10, max_length=1000)]
    resolution: Annotated[str, StringConstraints(min_length=10, max_length=1000)]
    customer_impact: Annotated[str, StringConstraints(min_length=10, max_length=800)]
    follow_ups: Annotated[
        list[Annotated[str, StringConstraints(min_length=5, max_length=320)]],
        Field(min_length=1, max_length=10),
    ]


class WorkflowEvent(ContractModel):
    id: Identifier
    sequence: Annotated[int, Field(gt=0)]
    timestamp: Timestamp
    type: WorkflowEventType
    phase: WorkflowPhase
    actor_id: Identifier
    actor_name: Identifier
    team: Team
    title: Headline
    detail: Annotated[str, StringConstraints(min_length=3, max_length=1200)]
    level: EventLevel


class WorkflowMetrics(ContractModel):
    agents_total: Annotated[int, Field(ge=0)]
    agents_completed: Annotated[int, Field(ge=0)]
    active_agents: Annotated[int, Field(ge=0)]
    tasks_completed: Annotated[int, Field(ge=0)]
    handoffs: Annotated[int, Field(ge=0)]
    confidence: Confidence


class WorkflowRun(ContractModel):
    id: Identifier
    incident: IncidentInput
    mode: ProviderMode
    status: WorkflowStatus
    phase: WorkflowPhase
    started_at: Timestamp
    updated_at: Timestamp
    agents: list[AgentRuntime]
    events: list[WorkflowEvent]
    findings: list[AgentFinding]
    team_reports: list[TeamReport]
    plan: RemediationPlan | None
    approval: ApprovalRequest | None
    verification: VerificationReport | None
    outcome: IncidentOutcome | None
    metrics: WorkflowMetrics


class ApprovalDecision(ContractModel):
    decision: Literal["approve", "reject"]
    reviewer: Annotated[str, StringConstraints(strip_whitespace=True, min_length=2, max_length=80)]
    note: Annotated[str, StringConstraints(strip_whitespace=True, max_length=500)] = ""


class ApiError(BaseModel):
    error: str
    details: list[str] | None = None


DEFAULT_INCIDENT = IncidentInput(
    title="Checkout latency surge across EU region",
    description=(
        "Checkout p95 latency climbed from 420 ms to 8.4 s shortly after the payment-router "
        "deployment. Error rate is 18% in eu-central and the retry queue continues to grow."
    ),
    service="payment-router",
    severity="SEV-1",
    region="eu-central",
    signals=[
        "p95 latency 8.4 s (baseline 420 ms)",
        "HTTP 5xx rate 18%",
        "payment retry queue 4.7x above baseline",
    ],
)
